typing/types.ml

```diff
@@ -478,19 +489,149 @@ let signature_item_id = function
+    | Heap, _ | _, Heap -> Heap
+    | Local, Local -> Local
+
+  exception NotSubmode
+(*
+  let pp_c ppf = function
+    | Heap -> Printf.fprintf ppf "0"
+    | Local -> Printf.fprintf ppf "1"
+  let pp_v ppf v =
+    let i = v.mvid in
+    (if i < 26 then Printf.fprintf ppf "%c" (Char.chr (Char.code 'a' + i))
+    else Printf.fprintf ppf "v%d" i);
+    Printf.fprintf ppf "[%a%a]" pp_c v.lower pp_c v.upper
+*)
+  let submode_cv m v =
+    (* Printf.printf "  %a <= %a\n" pp_c m pp_v v; *)
+    if le m v.lower then ()
+    else if not (le m v.upper) then raise NotSubmode
+    else begin
+      let m = join v.lower m in
+      v.lower <- m;
+      if m = v.upper then v.vlower <- []
+    end
+
+  let rec submode_vc v m =
+    (* Printf.printf "  %a <= %a\n" pp_v v pp_c m; *)
+    if le v.upper m then ()
+    else if not (le v.lower m) then raise NotSubmode
+    else begin
+      let m = meet v.upper m in
+      v.upper <- m;
+      v.vlower |> List.iter (fun a ->
+        (* a <= v <= m *)
+        submode_vc a m;
+        v.lower <- join v.lower a.lower;
+      );
+      if v.lower = m then v.vlower <- []
+    end
+
+  let submode_vv a b =
+    (* Printf.printf "  %a <= %a\n" pp_v a pp_v b; *)
+    if le a.upper b.lower then ()
+    else if List.memq a b.vlower then ()
+    else begin
+      submode_vc a b.upper;
+      b.vlower <- a :: b.vlower;
+      submode_cv a.lower b;
+    end
+
+  let submode a b =
+    match
+      match a, b with
+      | Amode a, Amode b ->
+         if not (le a b) then raise NotSubmode
+      | Amodevar v, Amode c ->
+         (* Printf.printf "%a <= %a\n" pp_v v pp_c c; *)
+         submode_vc v c
+      | Amode c, Amodevar v ->
+         (* Printf.printf "%a <= %a\n" pp_c c pp_v v; *)
+         submode_cv c v
+      | Amodevar a, Amodevar b ->
+         (* Printf.printf "%a <= %a\n" pp_v a pp_v b; *)
+         submode_vv a b
+    with
+    | () -> Ok ()
+    | exception NotSubmode -> Error ()
+
+  let next_id = ref (-1)
+  let fresh () =
+    incr next_id;
+    { upper = Local; lower = Heap; vlower = []; mvid = !next_id }
+
+  let joinvars vars =
+    let v = fresh () in
+    vars |> List.iter (fun v' -> submode_vv v' v);
+    v
 
   let join ms =
-    if List.mem Alloc_local ms then Alloc_local else Alloc_heap
+    let rec aux vars = function
+      | [] ->
+         if vars = [] then heap
+         else Amodevar (joinvars vars)
+      | Amode Heap :: ms -> aux vars ms
+      | Amode Local :: _ -> local
+      | Amodevar v :: ms -> aux (v :: vars) ms
+    in aux [] ms
+
+  let constrain_upper = function
+    | Amode m -> m
+    | Amodevar v ->
+       submode_cv v.upper v;
+       v.upper
+
+  let constrain_lower = function
+    | Amode m -> m
+    | Amodevar v when v.lower = v.upper -> v.lower
+    | Amodevar v ->
+       let rec trans vlow =
+         vlow |> List.iter (fun v' ->
+           if le v'.upper v.lower then ()
+           else if List.memq v' v.vlower then ()
```

Comment:
This looks like its going to make `trans` a no-op since you initially call it only on members of `v.vlower`.

Reply:
That's definitely wrong. Pushed a fix, does that look better?